torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp

```diff
@@ -703,6 +712,9 @@ class TORCH_API ProcessGroupNCCL : public Backend {
 
   std::list<ProcessGroupNCCL::WorkNCCL> completedWorkList_;
 
+  // All process groups for checking Watchdog status
+  static std::unordered_set<c10d::ProcessGroupNCCL*> all_nccl_process_groups;
+
```

Comment:
In case there are multiple threads and 1 GPU per thread, `all_nccl_process_groups` would include the PG objects on other threads. Not sure if this would create chance of data race (e.g. the return of `watchDogsDone` may depend on when _this_ thread check versus when the _other_ thread enqueues its collective.)

Reply:
The current solution is only intended to target the case where the work is enqueued by the main thread and captured on the main thread, as the failing case is work being enqueued during the _warmup_ on the main thread being queried during the capture itself. If work is enqueued on a side thread, then that is already unsafe as there would be no guarantee on what is captured/not captured, similar to Elias's concern.